include/continuable/detail/support/asio.hpp

```diff
@@ -0,0 +1,175 @@
+#define CTI_DETAIL_ASIO_HAS_NO_INTEGRATION
+#elif (BOOST_VERSION / 100 % 1000) <= 71
+#define CTI_DETAIL_ASIO_HAS_EXPLICIT_RET_TYPE_INTEGRATION
+#endif
+
+#if defined(CONTINUABLE_HAS_EXCEPTIONS)
+#include <boost/system/system_error.hpp>
+#endif
+
+#define CTI_DETAIL_ASIO_NAMESPACE_BEGIN                                        \
+  namespace boost {                                                            \
+  namespace asio {
+#define CTI_DETAIL_ASIO_NAMESPACE_END                                          \
+  }                                                                            \
+  }
+#endif
+
+#if defined(CTI_DETAIL_ASIO_HAS_NO_INTEGRATION)
+#error \
+  "First-class ASIO support for continuable requires the form of "\
+  "`async_result` with an `initiate` static member function, which was added " \
+  "in standalone ASIO 1.13.0 and Boost ASIO 1.70. Older versions can be " \
+  "integrated manually with `cti::promisify`."
+#endif
+
+#include <utility>
+
+namespace cti {
+namespace detail {
+namespace asio {
+
+#if defined(ASIO_STANDALONE)
+using error_code_t = ::asio::error_code;
+using error_cond_t = std::error_condition;
+
+#if defined(CONTINUABLE_HAS_EXCEPTIONS)
+using exception_t = ::asio::system_error;
+#endif
+#else
+using error_code_t = ::boost::system::error_code;
+using error_cond_t = ::boost::system::error_condition;
+
+#if defined(CONTINUABLE_HAS_EXCEPTIONS)
+using exception_t = ::boost::system::system_error;
+#endif
+#endif
+
+#if defined(CONTINUABLE_HAS_EXCEPTIONS)
+template <typename Promise>
+void promise_exception_helper(Promise& promise, error_code_t e) noexcept {
+  promise.set_exception(std::make_exception_ptr(exception_t(std::move(e))));
+}
+#else
+template <typename Promise>
+void promise_exception_helper(Promise& promise, error_code_t e) noexcept
+    ->decltype(promise.set_exception(std::move(e))) {
+  return promise.set_exception(std::move(e));
+}
+
+template <typename Promise>
+void promise_exception_helper(Promise& promise, error_code_t e) noexcept
+    ->decltype(promise.set_exception(error_cond_t(e.value(), e.category()))) {
+  promise.set_exception(error_cond_t(e.value(), e.category()));
+}
+
+#endif
+
+// Binds `promise` to the first argument of a continuable resolver, giving it
+// the signature of an ASIO handler.
+template <typename Promise>
+auto promise_resolver_handler(Promise&& promise) noexcept {
+  return [promise = std::forward<Promise>(promise)](
+             error_code_t e, auto&&... args) mutable noexcept {
+    if (e) {
+      promise_exception_helper(promise, std::move(e));
+    } else {
+      promise.set_value(std::forward<decltype(args)>(args)...);
+    }
+  };
+}
+
+// Helper struct wrapping a call to `cti::make_continuable` and, if needed,
+// providing an erased, explicit `return_type` for `async_result`.
+template <typename Signature>
+struct initiate_make_continuable;
+
+template <typename... Args>
+struct initiate_make_continuable<void(error_code_t, Args...)> {
```

Comment:
This can be specialized on the struct directly:

```cpp
template <typename Signature>
struct initiate_make_continuable;

template <>
struct initiate_make_continuable<void(error_code_t> {
  // ...
};

template <typename First, typename... Rest> // Probably also works witha single Args...
struct initiate_make_continuable<void(error_code_t, First, Rest...> {
  // ...
};
```

Reply:
I've updated to use `attorney::create_from` but I'm not sure if I follow the comment here. Now we have the following:

 ```c++
 template <typename Signature>
 struct initiate_make_continuable;
 
 template <typename... Args>
 struct initiate_make_continuable<void(error_code_t, Args...)> {
   // impl here, works regardless of sizeof...(Args)
 };
 
 template <typename... Args>
 struct initiate_make_continuable<void(error_code_t const&, Args...)>
  : initiate_make_continuable<void(error_code_t, Args...)> {
  // empty
 };
 ```

The key is that we allow users to specify async ops with handlers which take the error code by value or by reference to const (note `error_code` vs `error_code const&`), both of which are technically valid although `error_code const&` sees wider use